### PR TITLE
feat: repaint ui for setProperties in visual editor

### LINKED ISSUE #1701: Action nodes redesign to improve visual canvas' readability
### Description
Currently, all the action nodes have the same format(Label-sublabel) which works for few, but not for the others. This is demanding a redesign sooner or later.
The redesign's goal is to improve the overall readability of visual canvas through a flexible design system – you should be able to quickly scan through it and understand the flow, where to find what. 

### Execution plan
- [ ] (in PR #2005 ) Layout: smart layouter, support size-sensitive node blocks
- [ ] (in PR #2017 ) Widget: enhance the appearance of `SetProperties` as a multi-line node, extract shared elements
- [ ] Widget: implement multi-section node widget which is split by '-------v-------', apply to specific $types
- [ ] LG: Syntax hightlighting lib of LanguageGeneration(LG) text 
- [ ] Layout: node size adjustment and node styles fine tune

### Other context
Design link: https://www.figma.com/file/04DymSq4xElhOBdAyutvvTbq/Composer-SH-06.19?node-id=4585%3A1049

### PR DESCRIPTION
**Pull request by alanlong9278**
## Description
Repaint ui for `setProperties` with new design in visual editor.

### What does this pr do?
- Extract `listOverview` component.
Like the pics show,  the list in `setProperties` and `multiInput` is the same ui except  the itemRender.
So, we can extract `listOverview` and through sending renderFunciton to render different item list.
image
image

- Define the elements as styledComponents
Depends on new design, there are 4 types styled compoments. The element components can change the color by props.
1. Span 
image

2. SingleLineDiv
image

3. MultiLineDiv
image

4. BorderedDiv
image

4. LinkBtn
image

- Build `setProperties`, `ChoiceInput`  with elements.
- Use `LinkBtn`(elements) in `dialogRef` node.

### What's next?
- Add templateElement into element libs.
- Refactor `formCard` to support new design.(Ze is doing)

- Combine with the layout pr written by ze.
- Rewrite the node ui with new design.

## Task Item
fixes #1701

### CODE REVIEW

File: Composer/packages/extensions/visual-designer/src/components/common/ListOverview.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import { UIElement } from '../elements/styledComponents.types';
+
+export interface ListOverviewProps {
+  items: string[];
+  elementSchema: UIElement;
+  maxCount: number;
+  role: string;
+  styles?: Record<string, any>;
+}
+export const ListOverview: React.FC<ListOverviewProps> = ({ items, elementSchema, maxCount, role, styles }) => {
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/components/common/ListOverview.tsx]: As a standalone component, ListOverview needs to be independent of any schema-driven concept. So the prop `elementSchema: UIElement` here is a logic leak and needs to be refined.

You can make this component receive `React.Node / JSX.Element` as input `items`. For now let's just write those UIElement inline; ideally, `UIElement` could be handled by uischemaRenderer, we can do that later.

File: Composer/packages/extensions/visual-designer/src/widgets/ChoiceInput.tsx [hunk 2]
```diff
@@ -6,67 +6,45 @@ import { jsx } from '@emotion/core';
...
 import { getUserAnswersTitle } from '../components/nodes/utils';
+import { ChoiceInputSize, ChoiceInputMarginTop } from '../constants/ElementSizes';
+import { BorderedDiv } from '../components/elements/styledComponents';
 
 export const ChoiceInputChoices = ({ choices }) => {
   if (!Array.isArray(choices)) {
     return null;
   }
-
+  const itemRender = ({ children }) => (
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/widgets/ChoiceInput.tsx]: Capitalize it if it's a FC; or name it as `renderItem`.
```
const renderItem = (item: Item) => <></>
```

File: Composer/packages/extensions/visual-designer/src/widgets/ChoiceInput.tsx [hunk 3]
```diff
@@ -6,67 +6,45 @@ import { jsx } from '@emotion/core';
...
-            boxSizing: 'border-box',
-          }}
-        >
-          {`${choices.length - 3} more`}
-        </div>
-      ) : null}
+    <div data-testid="ChoiceInput" css={{ padding: '0 0 16px 8px' }}>
+      <ListOverview
+        items={choices.map(choice => choice.value)}
+        ItemRender={itemRender}
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/widgets/ChoiceInput.tsx]: props name might be `ItemRenderer` if passed in a FC;
or use `renderItem`

File: Composer/packages/extensions/visual-designer/src/widgets/PropertiesWidget.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      }}
+    >
+      {children}
+    </MultiLineDiv>
+  );
+  return (
+    <div data-testid="SetProperties" css={{ padding: '0 0 16px 8px' }}>
+      <ListOverview
+        items={items}
+        ItemRender={itemRender}
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/widgets/PropertiesWidget.tsx]: sync with ChoiceInput after changes

File: Composer/packages/extensions/visual-designer/src/widgets/PropertiesWidget.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+const DefaultCardColor = {
+  theme: ObiColors.AzureGray3,
+  icon: ObiColors.AzureGray2,
+};
+
+export const PropertiesWidget: WidgetComponent<PropertiesWidgetProps> = ({
+  id,
+  data,
+  onEvent,
+  title,
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/widgets/PropertiesWidget.tsx]: Sorry I made a confusion here. The `ChoiceInput` definition should be removed; you don't need to define a `PropertiesWidget` since its props is completely the same with `ActionCard`.

You can reference `uischema->line:35` to integrate your `AssignmentList` directly into uischema

**Comment by yeze322** [comment]
The styled component lib could help us standardize the design langauge we used in Visual Editor. It's a really good change.

We will migrate to the styled lib in future PRs.

File: Composer/packages/extensions/visual-designer/src/widgets/PropertiesWidget.tsx (hunk 5 above)
Comment (yeze322) [reply Composer/packages/extensions/visual-designer/src/widgets/PropertiesWidget.tsx]: But this is an **optional** changes. We can keep this widget because for future usage on `DeleteProperties`. Let's keep it, leave this comment FYI

Comment (alanlong9278) [reply Composer/packages/extensions/visual-designer/src/widgets/PropertiesWidget.tsx]: OK, thanks

File: Composer/packages/extensions/visual-designer/src/components/nodes/templates/FormCard.tsx [hunk 6]
```diff
@@ -95,47 +96,35 @@ export const FormCard: FunctionComponent<NodeProps> = ({
...
-                display: 'flex',
-                alignItems: 'center',
-                justifyContent: 'center',
-                marginRight: '5px',
-              }}
-            >
-              <Icon icon={icon} color={iconColor} size={iconSize || 16} />
-            </div>
-          )}
+        {label && (
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/components/nodes/templates/FormCard.tsx]: Is it necessary to change this file? There will be conflicts with another PR potentially.

Comment (alanlong9278) [reply Composer/packages/extensions/visual-designer/src/components/nodes/templates/FormCard.tsx]: Fixed conflict. This is beacause there is no lable in `setProperties` node. there is a empty div to take up space if without the condition.

File: Composer/packages/extensions/visual-designer/src/widgets/PropertiesWidget.tsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    icon: string;
+  };
+}
+
+const DefaultCardColor = {
+  theme: ObiColors.AzureGray3,
+  icon: ObiColors.AzureGray2,
+};
+
+export const PropertiesWidget: WidgetComponent<PropertiesWidgetProps> = ({
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/widgets/PropertiesWidget.tsx]: You can merge `PropertiesWidget` with `ListOverviewCard`, they have similar functionality

File: Composer/packages/extensions/visual-designer/src/components/nodes/templates/ListOverviewCard.tsx [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  corner?: any;
+  label?: any;
+  icon?: string;
+  iconSize?: number;
+  styles?: object;
+  nodeColors: { [key: string]: any };
+  onClick?: () => void;
+  children?: any;
+}
+export const ListOverviewCard: FunctionComponent<NodeProps> = ({
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/components/nodes/templates/ListOverviewCard.tsx]: `ListOverviewCard` should at least consume the `ListOverview` FC component, otherwise it's not a 'ListOverview' card.

File: Composer/packages/extensions/visual-designer/src/widgets/ActionCard.tsx [hunk 9]
```diff
@@ -47,8 +48,9 @@ export const ActionCard: WidgetComponent<ActionCardProps> = ({
 }) => {
   const header = disableSDKTitle ? title : generateSDKTitle(data, title);
   const nodeColors = { themeColor: colors.theme, iconColor: colors.icon };
+  const CardRender = data.$type === SDKTypes.ChoiceInput ? ListOverviewCard : FormCard;
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/widgets/ActionCard.tsx]: don't couple schema logic here. you can edit the uischema file

Comment (yeze322) [reply Composer/packages/extensions/visual-designer/src/widgets/ActionCard.tsx]: Now that we already have a new widget, ActionCard.tsx doesn't need to be touched

File: Composer/packages/extensions/visual-designer/src/widgets/ListOverviewWidget.tsx [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        <div css={{ padding: '0 0 8px 8px' }}>
+          <ListOverview
+            items={items}
+            ItemRender={ItemRender}
+            maxCount={3}
+            styles={{
+              height: itemSize.height,
+              width: itemSize.width,
+              marginTop: itemSize.marginTop,
+            }}
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/widgets/ListOverviewWidget.tsx]: item size should be part of ItemRenderer, in your case, should be binded to the styled component you passed in as `Component`.

Comment (yeze322) [reply Composer/packages/extensions/visual-designer/src/widgets/ListOverviewWidget.tsx]: The problem here is you passed in a pair of `weight` and `height`, intuitively, it should limit the whole list's size but actually it's for a single item.

File: Composer/packages/extensions/visual-designer/src/components/nodes/templates/ListOverviewCard.tsx (hunk 8 above)
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/components/nodes/templates/ListOverviewCard.tsx]: It's acceptable to not define a new ListOverviewCard even if text will be displayed as multiple lines.

The reason is 1. avoid making too much noise to other reviewers (100 lines of changes are scary) 2. the name `ListOverviewCard` is not self-described, as I said, it doesn't contain any ListOverview logic.

File: Composer/packages/extensions/visual-designer/src/components/elements/styledComponents.tsx [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+}));
+
+export const Span = styled.span`
+  ${dynamicStyle}
+`;
+
+export const BorderedDiv: ElementComponent<ElementProps> = styled.div<ElementProps>(props => ({
+  color: props.color || ObiColors.Black,
+  width: ChoiceInputSize.width,
+  height: ChoiceInputSize.height,
```
[the remark this answers is not in the document]
Comment (yeze322) [reply Composer/packages/extensions/visual-designer/src/components/elements/styledComponents.tsx]: To the condition you need a specific size, just invoke is as `<BorderedDiv height={14} />`

File: Composer/packages/extensions/visual-designer/src/components/common/ListOverview.tsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+  items: string[];
+  ItemRender: FC<any> | ComponentClass<any, any>;
+  maxCount: number;
+  styles?: Record<string, any>;
+}
+export const ListOverview: React.FC<ListOverviewProps> = ({ items, ItemRender, maxCount, styles }) => {
+  return (
+    <>
+      {items.map((value, index) => {
+        if (index < 3) {
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/components/common/ListOverview.tsx]: use slice() instead checking index insde a loop

File: Composer/packages/extensions/visual-designer/src/components/common/ListOverview.tsx [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+          return (
+            <ItemRender key={index} title={typeof value === 'string' ? value : ''}>
+              {value}
+            </ItemRender>
+          );
+        }
+      })}
+      {items.length > maxCount ? (
+        <MultiLineDiv
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/components/common/ListOverview.tsx]: Semantically a center-aligned text. Just give it some styles instead of receiving from props. Its style should be a part of the component logic (exposure v.s. encapsulation)

File: Composer/packages/extensions/visual-designer/src/components/common/ListOverview.tsx [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+      })}
+      {items.length > maxCount ? (
+        <SingleLineDiv
+          data-testid="hasMore"
+          style={{
+            marginTop: '8px',
+            textAlign: 'center',
+          }}
+        >
+          {`${items.length - 3} more`}
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/components/common/ListOverview.tsx]: found a bug here: replace 3 with maxCount

File: Composer/packages/extensions/visual-designer/src/components/common/ListOverview.tsx [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+export interface ListOverviewProps {
+  items: string[];
+  ItemRender: FC<any> | ComponentClass<any, any>;
+  maxCount: number;
+  styles?: Record<string, any>;
+}
+export const ListOverview: FC<ListOverviewProps> = ({ items, ItemRender, maxCount, styles }) => {
+  return (
+    <div css={{ padding: '0 0 8px 8px' }}>
+      {items.slice(0, 3).map((value, index) => {
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/components/common/ListOverview.tsx]: hardcoded number '3'

File: Composer/packages/extensions/visual-designer/src/schema/uischema.tsx [hunk 16]
```diff
@@ -40,14 +42,34 @@ const BaseInputSchema: UIWidget = {
...
+    icon: ElementIcon.User,
+    menu: 'none',
+    content: data => <SingleLineDiv>{data.property || '<property>'}</SingleLineDiv>,
+    children: data => {
+      const items = Array.isArray(data.choices) ? data.choices.map(choice => choice.value) : [];
+      const ItemRender = props => (
+        <BorderedDiv width={ChoiceInputSize.width} height={ChoiceInputSize.height} {...props} />
+      );
+      return <ListOverview items={items} ItemRender={ItemRender} maxCount={3} />;
+    },
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/schema/uischema.tsx]: one last change, simplify the code inside uischema.
```
children: data =>
    <ListOverview
        items={data.assignments}
        renderItem={
            item => <BorderedDiv>{item.value}</BorderedDiv>
        }
    />
```

Comment (yeze322) [reply Composer/packages/extensions/visual-designer/src/schema/uischema.tsx]: it requires you optimize the ListOverview component

File: Composer/packages/extensions/visual-designer/src/schema/uischema.tsx [hunk 17]
```diff
@@ -127,6 +149,20 @@ export const uiSchema: UISchema = {
...
+        : [];
+      const ItemRender = props => (
+        <SingleLineDiv
+          width={PropertyAssignmentSize.width}
+          height={PropertyAssignmentSize.height}
+          style={{ paddingLeft: '3px' }}
+          {...props}
+        />
+      );
+      return <ListOverview items={items} ItemRender={ItemRender} maxCount={3} />;
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/schema/uischema.tsx]: ditto, shortten the code by optimizing the ListOverview component

File: Composer/packages/extensions/visual-designer/src/widgets/ActionCard.tsx [hunk 18]
```diff
@@ -16,6 +16,7 @@ export interface ActionCardProps extends WidgetContainerProps {
   content: string | number | JSX.Element;
   menu?: JSX.Element | string;
   children?: JSX.Element;
+  template?: JSX.Element;
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/widgets/ActionCard.tsx]: clean up

File: Composer/packages/extensions/visual-designer/src/components/common/ListOverview.tsx [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+import { jsx } from '@emotion/core';
+import { FC, ComponentClass } from 'react';
+
+import { SingleLineDiv } from '../elements/styledComponents';
+
+export interface ListOverviewProps {
+  items: string[];
+  ItemRender: FC<any> | ComponentClass<any, any>;
+  maxCount: number;
+  styles?: Record<string, any>;
```
Comment (yeze322) [inline Composer/packages/extensions/visual-designer/src/components/common/ListOverview.tsx]: React.CSSProperties
